File: story_monitor.py
import hashlib
import random
import re
import time
from pathlib import Path
from typing import Dict, Any, Optional
from playwright.sync_api import sync_playwright

from config import (
    SELLERS,
    DB_PATH,
    SESSION_DIR,
    HEADLESS,
    MAX_STORIES_PER_SELLER,
    CAR_IMAGES_DIR,
    SELLER_COOLDOWN_MIN,
    SELLER_COOLDOWN_MAX,
)
from database import (
    init_db,
    car_exists,
    save_car,
    get_db_summary,
    record_daily_check,
    get_daily_checks_summary,
    get_last_story_id,
    update_last_story_id,
)
from parsers import parse_car_text, is_valid_listing
from ocr_service import ocr_service
from image_service import (
    download_image,
    save_story_screenshot,
    get_story_media_image_url,
    create_temp_screenshot,
    clean_temp_file,
    clean_all_temp_files,
    safe_filename,
)
from instagram_client import (
    wait_for_login,
    get_story_fingerprint,
    find_post_in_story,
    go_to_next_story,
    get_post_info,
)


def format_car_summary_line(car_data: Dict[str, Any]) -> str:
    brand = car_data.get("brand") or ""
    model = car_data.get("model") or ""
    year = f" {car_data['year']}" if car_data.get("year") else ""
    return f"🚗 {brand} {model}{year}".strip()


def format_price_line(car_data: Dict[str, Any]) -> str:
    prices = []
    if car_data.get("price_tjs"):
        prices.append(f"{car_data['price_tjs']} TJS")
    if car_data.get("price_usd"):
        prices.append(f"{car_data['price_usd']} USD")
    return f"💰 {' / '.join(prices)}" if prices else "💰 No price specified"


def extract_story_id_from_url(url: str) -> Optional[str]:
    """Extracts numeric story ID from Instagram story URL."""
    if not url:
        return None
    match = re.search(r"/stories/[^/]+/(\d+)", url)
    return match.group(1) if match else None


def is_story_id_older_or_equal(current_id: Optional[str], checkpoint_id: Optional[str]) -> bool:
    """Checks if current story ID has already been seen in a previous run."""
    if not current_id or not checkpoint_id:
        return False

    try:
        return int(current_id) <= int(checkpoint_id)
    except ValueError:
        return str(current_id) == str(checkpoint_id)


def check_seller(context, page, seller: str, stats: Dict[str, int]):
    print("\n=========================================")
    print(f"Checking @{seller}")

    story_url = f"https://www.instagram.com/stories/{seller}/"

    try:
        page.goto(
            story_url,
            wait_until="domcontentloaded",
            timeout=60_000,
        )
    except Exception as error:
        print(f"❌ Failed to open stories: {error}")
        stats["errors"] += 1
        record_daily_check(seller, 0, 0)
        return

    page.wait_for_timeout(3000)

    if "/stories/" not in page.url:
        print(f"ℹ️ No active stories found for @{seller}")
        record_daily_check(seller, 0, 0)
        return

    checkpoint_story_id = get_last_story_id(seller)
    if checkpoint_story_id:
        print(f"📌 Last checkpoint Story ID: {checkpoint_story_id}")

    highest_story_id_seen = checkpoint_story_id
    visited_fingerprints = set()
    seller_stories_checked = 0
    seller_cars_found = 0

    for index in range(MAX_STORIES_PER_SELLER):
        print(f"\nSTORY #{index + 1}")
        stats["stories_checked"] += 1
        seller_stories_checked += 1

        fingerprint = get_story_fingerprint(page)
        if fingerprint in visited_fingerprints:
            print("Duplicate story detected. Moving to next seller.")
            break
        visited_fingerprints.add(fingerprint)

        current_story_id = extract_story_id_from_url(page.url)

        # ---------------------------------------------------------
        # FAST CHECKPOINT SKIP: If this story was already processed in previous run
        # ---------------------------------------------------------
        if current_story_id and checkpoint_story_id and is_story_id_older_or_equal(current_story_id, checkpoint_story_id):
            print(f"⏭ Story ID {current_story_id} <= checkpoint ({checkpoint_story_id}). FAST SKIP.")
            stats["duplicates_skipped"] += 1
            changed = go_to_next_story(page)
            if not changed or "/stories/" not in page.url:
                print(f"✅ Finished stories for @{seller}.")
                break
            continue

        # Track the highest numeric story ID seen in this run
        if current_story_id:
            if not highest_story_id_seen:
                highest_story_id_seen = current_story_id
            else:
                try:
                    if int(current_story_id) > int(highest_story_id_seen):
                        highest_story_id_seen = current_story_id
                except ValueError:
                    highest_story_id_seen = current_story_id

        post_url = find_post_in_story(page)

        # =========================================================
        # VARIANT 1: POST / REEL SHARE
        # =========================================================
        if post_url:
            source_key = f"POST|{seller}|{post_url}"
            print(f"🔗 POST: {post_url}")

            if car_exists(source_key):
                print("⏭ Already saved. SKIP.")
                stats["duplicates_skipped"] += 1
            else:
                try:
                    caption, image_url = get_post_info(context, post_url)
                    car_data = parse_car_text(caption)

                    if is_valid_listing(car_data):
                        image_path = download_image(context, image_url, seller, source_key)
                        if not image_path:
                            image_path = save_story_screenshot(page, seller, source_key)

                        save_car(
                            seller_username=seller,
                            source_type="POST_CAPTION",
                            source_key=source_key,
                            source_url=post_url,
                            raw_text=caption,
                            car_data=car_data,
                            image_url=image_url,
                            image_path=image_path,
                        )

                        print(format_car_summary_line(car_data))
                        print(format_price_line(car_data))
                        if car_data.get("phone_number"):
                            print(f"📞 Phone: {car_data['phone_number']}")
                        if image_path:
                            print(f"🖼 Image saved: {image_path}")
                        print("✅ Saved to DB")
                        stats["cars_added"] += 1
                        seller_cars_found += 1
                    else:
                        print("ℹ️ Model or price not found. SKIP.")
                        stats["invalid_skipped"] += 1
                except Exception as e:
                    print(f"⚠️ Post parse error: {e}")
                    stats["errors"] += 1

        # =========================================================
        # VARIANT 2: STORY OCR (DIRECT STORY TEXT)
        # =========================================================
        else:
            temp_screenshot = create_temp_screenshot(page, seller, index)
            try:
                lines = ocr_service.extract_text_from_image(temp_screenshot)
                raw_text = "\n".join(lines)

                print("📸 STORY OCR")
                print("TEXT:")
                print(raw_text if raw_text.strip() else "(no text detected)")

                car_data = parse_car_text(raw_text)

                # Generate unique key for story text
                clean_ocr = re.sub(r"\W+", "", raw_text.lower())
                if not clean_ocr:
                    clean_ocr = fingerprint

                story_hash = hashlib.sha256(f"{seller}|{clean_ocr}".encode("utf-8")).hexdigest()[:24]
                source_key = f"STORY|{seller}|{story_hash}"

                if car_exists(source_key):
                    print("⏭ Already saved. SKIP.")
                    stats["duplicates_skipped"] += 1
                    clean_temp_file(temp_screenshot)

                elif is_valid_listing(car_data):
                    media_url = get_story_media_image_url(page)
                    image_path = download_image(context, media_url, seller, source_key)

                    # If original media not downloadable, retain OCR screenshot as car image
                    if not image_path and temp_screenshot.exists():
                        seller_dir = CAR_IMAGES_DIR / safe_filename(seller)
                        seller_dir.mkdir(parents=True, exist_ok=True)
                        target_file = seller_dir / f"{story_hash[:16]}_story.png"
                        try:
                            temp_screenshot.replace(target_file)
                            image_path = str(target_file)
                        except Exception:
                            image_path = str(temp_screenshot)
                    else:
                        clean_temp_file(temp_screenshot)

                    save_car(
                        seller_username=seller,
                        source_type="STORY_OCR",
                        source_key=source_key,
                        source_url=page.url,
                        raw_text=raw_text,
                        car_data=car_data,
                        image_url=media_url,
                        image_path=image_path,
                    )

                    print(f"\n{format_car_summary_line(car_data)}")
                    print(format_price_line(car_data))
                    if car_data.get("phone_number"):
                        print(f"📞 Phone: {car_data['phone_number']}")
                    if image_path:
                        print(f"🖼 Image saved: {image_path}")
                    print("✅ Saved to DB")
                    stats["cars_added"] += 1
                    seller_cars_found += 1

                else:
                    print("ℹ️ Model or price not found. SKIP.")
                    stats["invalid_skipped"] += 1
                    clean_temp_file(temp_screenshot)

            except Exception as e:
                print(f"⚠️ Story OCR error: {e}")
                stats["errors"] += 1
                clean_temp_file(temp_screenshot)

        # Move to next slide
        changed = go_to_next_story(page)
        if not changed or "/stories/" not in page.url:
            print(f"✅ Finished stories for @{seller}.")
            break

    # Save highest seen story ID checkpoint for this seller
    if highest_story_id_seen:
        update_last_story_id(seller, highest_story_id_seen)

    # Record daily check metrics for this seller
    record_daily_check(seller, seller_stories_checked, seller_cars_found)


def main():
    init_db()

    # Pre-warm PaddleOCR
    _ = ocr_service.ocr

    print(f"Total sellers: {len(SELLERS)}")
    print(f"Database ready: {DB_PATH}")
    print(f"Images folder: {CAR_IMAGES_DIR}")

    stats = {
        "sellers_checked": 0,
        "stories_checked": 0,
        "cars_added": 0,
        "duplicates_skipped": 0,
        "invalid_skipped": 0,
        "errors": 0,
    }

    with sync_playwright() as p:
        print("\nLaunching browser...")

        context = p.chromium.launch_persistent_context(
            user_data_dir=str(SESSION_DIR),
            headless=HEADLESS,
            viewport={"width": 1400, "height": 1000},
            args=["--disable-notifications"],
        )

        page = context.pages[0] if context.pages else context.new_page()

        if not wait_for_login(context, page):
            context.close()
            return

        for index, seller in enumerate(SELLERS):
            stats["sellers_checked"] += 1
            try:
                check_seller(context, page, seller, stats)
            except Exception as error:
                print(f"\n❌ Unexpected error while checking @{seller}:")
                print(error)
                stats["errors"] += 1

            # Apply random anti-bot cooldown delay before next seller
            if index < len(SELLERS) - 1:
                cooldown = random.uniform(SELLER_COOLDOWN_MIN, SELLER_COOLDOWN_MAX)
                time.sleep(cooldown)

        context.close()

    # Cleanup temporary images
    clean_all_temp_files()

    # Final Summary Report
    print("\n=========================================")
    print("MONITORING COMPLETE\n")
    print(f"Sellers checked:    {stats['sellers_checked']}")
    print(f"Stories checked:    {stats['stories_checked']}")
    print(f"Cars added:         {stats['cars_added']}")
    print(f"Duplicates skipped: {stats['duplicates_skipped']}")
    print(f"Invalid skipped:    {stats['invalid_skipped']}")
    print(f"Errors:             {stats['errors']}")
    print(f"\nDatabase: {DB_PATH}")
    print(f"Images: {CAR_IMAGES_DIR}")

    summary = get_db_summary()
    print(f"\n🚗 Total cars in DB: {summary['total']} ({summary['with_phone']} with phone number)")
    for seller, count in summary["breakdown"].items():
        print(f"  @{seller}: {count}")

    # Display today's daily story check history
    daily_checks = get_daily_checks_summary()
    if daily_checks:
        print("\n📅 Today's Story Check Activity:")
        for check in daily_checks:
            print(f"  @{check['seller_username']}: {check['stories_count']} stories checked, {check['cars_found']} cars found (Last: {check['last_checked_at']})")
    print("=========================================\n")


if __name__ == "__main__":
    main()

File: src/storyMonitor.ts
import crypto from 'node:crypto';
import path from 'node:path';
import fs from 'node:fs';
import { chromium, type BrowserContext, type Page } from 'playwright';

import {
  SELLERS,
  DB_PATH,
  SESSION_DIR,
  HEADLESS,
  MAX_STORIES_PER_SELLER,
  CAR_IMAGES_DIR,
  SELLER_COOLDOWN_MIN,
  SELLER_COOLDOWN_MAX,
} from './config.js';
import {
  initDb,
  carExists,
  saveCar,
  getDbSummary,
  recordDailyCheck,
  getDailyChecksSummary,
  getLastStoryId,
  updateLastStoryId,
} from './database.js';
import { parseCarText, isValidListing } from './parsers.js';
import { ocrService } from './ocrService.js';
import {
  downloadImage,
  saveStoryScreenshot,
  getStoryMediaImageUrl,
  createTempScreenshot,
  cleanTempFile,
  cleanAllTempFiles,
  safeFilename,
} from './imageService.js';
import {
  waitForLogin,
  getStoryFingerprint,
  findPostInStory,
  goToNextStory,
  getPostInfo,
} from './instagramClient.js';
import type { CarParseResult, MonitoringStats } from './types.js';

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function randomBetween(min: number, max: number): number {
  return Math.random() * (max - min) + min;
}

function formatCarSummaryLine(carData: CarParseResult): string {
  const brand = carData.brand || '';
  const model = carData.model || '';
  const year = carData.year ? ` ${carData.year}` : '';
  return `🚗 ${brand} ${model}${year}`.trim();
}

function formatPriceLine(carData: CarParseResult): string {
  const prices: string[] = [];
  if (carData.price_tjs) prices.push(`${carData.price_tjs.toLocaleString('ru-RU')} TJS`);
  if (carData.price_usd) prices.push(`${carData.price_usd.toLocaleString('ru-RU')} USD`);
  return prices.length > 0 ? `💰 ${prices.join(' / ')}` : '💰 No price specified';
}

function extractStoryIdFromUrl(url: string): string | null {
  if (!url) return null;
  const match = url.match(/\/stories\/[^/]+\/(\d+)/);
  return match ? match[1] : null;
}

function isStoryIdOlderOrEqual(
  currentId: string | null,
  checkpointId: string | null
): boolean {
  if (!currentId || !checkpointId) return false;
  try {
    return BigInt(currentId) <= BigInt(checkpointId);
  } catch {
    return currentId === checkpointId;
  }
}

async function checkSeller(
  context: BrowserContext,
  page: Page,
  seller: string,
  stats: MonitoringStats
): Promise<void> {
  console.log('\n=========================================');
  console.log(`Checking @${seller}`);

  const storyUrl = `https://www.instagram.com/stories/${seller}/`;

  try {
    await page.goto(storyUrl, {
      waitUntil: 'domcontentloaded',
      timeout: 60000,
    });
  } catch (error) {
    console.log(`❌ Failed to open stories:`, error);
    stats.errors += 1;
    recordDailyCheck(seller, 0, 0);
    return;
  }

  await page.waitForTimeout(3000);

  if (!page.url().includes('/stories/')) {
    console.log(`ℹ️ No active stories found for @${seller}`);
    recordDailyCheck(seller, 0, 0);
    return;
  }

  const checkpointStoryId = getLastStoryId(seller);
  if (checkpointStoryId) {
    console.log(`📌 Last checkpoint Story ID: ${checkpointStoryId}`);
  }

  let highestStoryIdSeen = checkpointStoryId;
  const visitedFingerprints = new Set<string>();
  let sellerStoriesChecked = 0;
  let sellerCarsFound = 0;

  for (let index = 0; index < MAX_STORIES_PER_SELLER; index++) {
    console.log(`\nSTORY #${index + 1}`);
    stats.stories_checked += 1;
    sellerStoriesChecked += 1;

    const fingerprint = await getStoryFingerprint(page);
    if (visitedFingerprints.has(fingerprint)) {
      console.log('Duplicate story detected. Moving to next seller.');
      break;
    }
    visitedFingerprints.add(fingerprint);

    const currentStoryId = extractStoryIdFromUrl(page.url());

    // FAST CHECKPOINT SKIP
    if (
      currentStoryId &&
      checkpointStoryId &&
      isStoryIdOlderOrEqual(currentStoryId, checkpointStoryId)
    ) {
      console.log(
        `⏭ Story ID ${currentStoryId} <= checkpoint (${checkpointStoryId}). FAST SKIP.`
      );
      stats.duplicates_skipped += 1;
      const changed = await goToNextStory(page);
      if (!changed || !page.url().includes('/stories/')) {
        console.log(`✅ Finished stories for @${seller}.`);
        break;
      }
      continue;
    }

    // Track highest story ID
    if (currentStoryId) {
      if (!highestStoryIdSeen) {
        highestStoryIdSeen = currentStoryId;
      } else {
        try {
          if (BigInt(currentStoryId) > BigInt(highestStoryIdSeen)) {
            highestStoryIdSeen = currentStoryId;
          }
        } catch {
          highestStoryIdSeen = currentStoryId;
        }
      }
    }

    const postUrl = await findPostInStory(page);

    // =========================================================
    // VARIANT 1: POST / REEL SHARE
    // =========================================================
    if (postUrl) {
      const sourceKey = `POST|${seller}|${postUrl}`;
      console.log(`🔗 POST: ${postUrl}`);

      if (carExists(sourceKey)) {
        console.log('⏭ Already saved. SKIP.');
        stats.duplicates_skipped += 1;
      } else {
        try {
          const [caption, imageUrl] = await getPostInfo(context, postUrl);
          const carData = parseCarText(caption);

          if (isValidListing(carData)) {
            let imagePath = await downloadImage(context, imageUrl, seller, sourceKey);
            if (!imagePath) {
              imagePath = await saveStoryScreenshot(page, seller, sourceKey);
            }

            saveCar(
              seller,
              'POST_CAPTION',
              sourceKey,
              postUrl,
              carData,
              imageUrl,
              imagePath
            );

            console.log(formatCarSummaryLine(carData));
            console.log(formatPriceLine(carData));
            if (carData.phone_number) {
              console.log(`📞 Phone: ${carData.phone_number}`);
            }
            if (imagePath) {
              console.log(`🖼 Image saved: ${imagePath}`);
            }
            console.log('✅ Saved to DB');
            stats.cars_added += 1;
            sellerCarsFound += 1;
          } else {
            console.log('ℹ️ Model or price not found. SKIP.');
            stats.invalid_skipped += 1;
          }
        } catch (e) {
          console.warn(`⚠️ Post parse error:`, e);
          stats.errors += 1;
        }
      }
    }

    // =========================================================
    // VARIANT 2: STORY OCR
    // =========================================================
    else {
      const tempScreenshot = await createTempScreenshot(page, seller, index);
      try {
        const lines = await ocrService.extractTextFromImage(tempScreenshot);
        const rawText = lines.join('\n');

        console.log('📸 STORY OCR');
        console.log('TEXT:');
        console.log(rawText.trim() ? rawText : '(no text detected)');

        const carData = parseCarText(rawText);

        const cleanOcr = rawText.toLowerCase().replace(/\W+/g, '') || fingerprint;
        const storyHash = crypto
          .createHash('sha256')
          .update(`${seller}|${cleanOcr}`)
          .digest('hex')
          .slice(0, 24);
        const sourceKey = `STORY|${seller}|${storyHash}`;

        if (carExists(sourceKey)) {
          console.log('⏭ Already saved. SKIP.');
          stats.duplicates_skipped += 1;
          cleanTempFile(tempScreenshot);
        } else if (isValidListing(carData)) {
          const mediaUrl = await getStoryMediaImageUrl(page);
          let imagePath = await downloadImage(context, mediaUrl, seller, sourceKey);

          if (!imagePath && fs.existsSync(tempScreenshot)) {
            const sellerDir = path.join(CAR_IMAGES_DIR, safeFilename(seller));
            fs.mkdirSync(sellerDir, { recursive: true });
            const targetFile = path.join(sellerDir, `${storyHash.slice(0, 16)}_story.png`);
            try {
              fs.renameSync(tempScreenshot, targetFile);
              imagePath = `car_images/${safeFilename(seller)}/${storyHash.slice(0, 16)}_story.png`;
            } catch {
              imagePath = tempScreenshot;
            }
          } else {
            cleanTempFile(tempScreenshot);
          }

          saveCar(
            seller,
            'STORY_OCR',
            sourceKey,
            page.url(),
            carData,
            mediaUrl,
            imagePath
          );

          console.log(`\n${formatCarSummaryLine(carData)}`);
          console.log(formatPriceLine(carData));
          if (carData.phone_number) {
            console.log(`📞 Phone: ${carData.phone_number}`);
          }
          if (imagePath) {
            console.log(`🖼 Image saved: ${imagePath}`);
          }
          console.log('✅ Saved to DB');
          stats.cars_added += 1;
          sellerCarsFound += 1;
        } else {
          console.log('ℹ️ Model or price not found. SKIP.');
          stats.invalid_skipped += 1;
          cleanTempFile(tempScreenshot);
        }
      } catch (e) {
        console.warn(`⚠️ Story OCR error:`, e);
        stats.errors += 1;
        cleanTempFile(tempScreenshot);
      }
    }

    // Move to next slide
    const changed = await goToNextStory(page);
    if (!changed || !page.url().includes('/stories/')) {
      console.log(`✅ Finished stories for @${seller}.`);
      break;
    }
  }

  // Update last seen story checkpoint
  if (highestStoryIdSeen) {
    updateLastStoryId(seller, highestStoryIdSeen);
  }

  // Record daily check
  recordDailyCheck(seller, sellerStoriesChecked, sellerCarsFound);
}

export async function main(): Promise<void> {
  initDb();

  console.log(`Total sellers: ${SELLERS.length}`);
  console.log(`Database ready: ${DB_PATH}`);
  console.log(`Images folder: ${CAR_IMAGES_DIR}`);

  const stats: MonitoringStats = {
    sellers_checked: 0,
    stories_checked: 0,
    cars_added: 0,
    duplicates_skipped: 0,
    invalid_skipped: 0,
    errors: 0,
  };

  console.log('\nLaunching browser...');

  const context = await chromium.launchPersistentContext(SESSION_DIR, {
    headless: HEADLESS,
    viewport: { width: 1400, height: 1000 },
    args: ['--disable-notifications'],
  });

  const page = context.pages()[0] || (await context.newPage());

  const loggedIn = await waitForLogin(context, page);
  if (!loggedIn) {
    await context.close();
    return;
  }

  for (let index = 0; index < SELLERS.length; index++) {
    const seller = SELLERS[index];
    stats.sellers_checked += 1;

    try {
      await checkSeller(context, page, seller, stats);
    } catch (error) {
      console.log(`\n❌ Unexpected error while checking @${seller}:`, error);
      stats.errors += 1;
    }

    if (index < SELLERS.length - 1) {
      const cooldown = randomBetween(SELLER_COOLDOWN_MIN, SELLER_COOLDOWN_MAX);
      await sleep(cooldown * 1000);
    }
  }

  await context.close();
  await ocrService.terminate();
  cleanAllTempFiles();

  // Summary Report
  console.log('\n=========================================');
  console.log('MONITORING COMPLETE\n');
  console.log(`Sellers checked:    ${stats.sellers_checked}`);
  console.log(`Stories checked:    ${stats.stories_checked}`);
  console.log(`Cars added:         ${stats.cars_added}`);
  console.log(`Duplicates skipped: ${stats.duplicates_skipped}`);
  console.log(`Invalid skipped:    ${stats.invalid_skipped}`);
  console.log(`Errors:             ${stats.errors}`);
  console.log(`\nDatabase: ${DB_PATH}`);
  console.log(`Images: ${CAR_IMAGES_DIR}`);

  const summary = getDbSummary();
  console.log(
    `\n🚗 Total cars in DB: ${summary.total} (${summary.with_phone} with phone number)`
  );
  for (const [seller, count] of Object.entries(summary.breakdown)) {
    console.log(`  @${seller}: ${count}`);
  }

  const dailyChecks = getDailyChecksSummary();
  if (dailyChecks.length > 0) {
    console.log("\n📅 Today's Story Check Activity:");
    for (const check of dailyChecks) {
      console.log(
        `  @${check.seller_username}: ${check.stories_count} stories checked, ${check.cars_found} cars found (Last: ${check.last_checked_at})`
      );
    }
  }
  console.log('=========================================\n');
}

// Auto-run if executed directly
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => {
    console.error('Fatal execution error:', err);
    process.exit(1);
  });
}
